File: src/FlashcardParser/SingleLineParser/Card.ts
import { IFlashcard, IStudySet, LineDescriptor } from "../Types/Types";
import { PageRef } from "@/Commands/All/PageRef";
import { Header } from "@/Commands/All/Header";
import { parseCommandLine } from "./Command";

export function parseCardLine(
	lineDescriptor: LineDescriptor,
	studySet: IStudySet
): boolean {
	const { trimmedLine: line } = lineDescriptor;

	let cardFront = line;
	const parts = line.split(PageRef.symbol);
	if (parts.length == 2) {
		parseCommandLine(lineDescriptor, studySet, `${PageRef.symbol}${parts[1]}`, true);
		cardFront = parts[0].trim();
	}

	const card: IFlashcard = {
		lineDescriptor: lineDescriptor,
		text: cardFront,
		headers: [],
		// FSRS initial difficulty and stability
		ease: 5,
		interval: 0.5,
		learningPhase: true,
		reviewedAt: null,
		retrievalSuccess: 0,
		reviewCount: 0,
		nextReviewAt: new Date(), // Default to now, will be updated if recall data follows
		skippedUntil: null,
	};

	addHeaders(card, studySet);

	studySet.flashcards.push(card);

	return true;
}

/**
 * Adds to the card's headers field the list of all heards inside which the card
 * is contained.
 * @param card
 * @param studySet
 */
function addHeaders(card: IFlashcard, studySet: IStudySet) {
	let j = studySet.headers.length - 1;
	let lastHeaderLevel: number | undefined;

	while (j >= 0) {
		const lineDescriptor = studySet.headers[j];
		j--;

		if (lineDescriptor.index > card.lineDescriptor.index) continue;

		const header = lineDescriptor.subParts.find((x) => x.name === Header.commandName);
		if (lastHeaderLevel !== undefined && header.level >= lastHeaderLevel) {
			break;
		}

		card.headers.unshift(header.text); // Add to beginning to maintain order
		lastHeaderLevel = header.level;
	}
}

File: service/flashcard_parser/single_line_parser/card.py
from typing import List
from datetime import datetime

from ..types import Flashcard, StudySet, LineDescriptor
from ..commands.page_ref import PageRef
from .command import parse_command_line


def parse_card_line(line_descriptor: LineDescriptor, study_set: StudySet) -> bool:
    line = line_descriptor.trimmed_line
    card_front = line
    parts = line.split(PageRef.symbol)
    if len(parts) == 2:
        parse_command_line(line_descriptor, study_set, f"{PageRef.symbol}{parts[1]}", True)
        card_front = parts[0].strip()

    card = Flashcard(
        line_descriptor=line_descriptor,
        text=card_front,
        headers=[],
        ease=5,
        interval=0.5,
        learning_phase=True,
        reviewed_at=None,
        retrieval_success=0,
        review_count=0,
        next_review_at=datetime.utcnow(),
    )

    _add_headers(card, study_set)
    study_set.flashcards.append(card)
    return True


def _add_headers(card: Flashcard, study_set: StudySet) -> None:
    j = len(study_set.headers) - 1
    last_header_level = None
    while j >= 0:
        ld = study_set.headers[j]
        j -= 1
        if ld.index > card.line_descriptor.index:
            continue
        header = next((x for x in ld.sub_parts if x.get("name") == "Header"), None)
        if header is None:
            continue
        if last_header_level is not None and header.get("level") >= last_header_level:
            break
        card.headers.insert(0, header.get("text"))
        last_header_level = header.get("level")
